### PR TITLE
[FIx #9227] Change to better suggestion for Style/GuardClause

### LINKED ISSUE #9227: Misleading example for Style/GuardClause
The example in the documentation for `Style/GuardClause` is misleading because mentioned bad and good version don#t match execution.

It says

```
# bad
if something
  foo || raise('exception')
else
  ok
end

# good
foo || raise('exception') if something
ok
```
However both versions behave differently and _bad_ can't be replaced by _good_.

Imagine `something` being `true` and `foo` returning a truthy value.
Bad (original) behavior: `foo` gets called.
Good (suggested) behavior: `foo` gets called and `ok` gets called.

I suggest to remove this example because I don't see a new case that would work which hasn't been mentioned before.

### PR DESCRIPTION
**Pull request by KessaPassa**
**Replace this text with a summary of the changes in your PR.
The more detailed you are, the better.**

follow up to #10868

-----------------

Before submitting the PR make sure the following are checked:

* [ ] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [ ] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [ ] Feature branch is up-to-date with `master` (if not - rebase it).
* [ ] Squashed related commits together.
* [ ] Added tests.
* [ ] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [ ] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/guard_clause.rb [hunk 1]
```diff
@@ -157,11 +157,13 @@ def register_offense(node, scope_exiting_keyword, conditional_keyword)
...
 
-        def guard_clause_source(guard_clause)
+        def guard_clause_exiting_keyword(guard_clause)
           parent = guard_clause.parent
 
-          if parent.and_type? || parent.or_type?
+          if parent.and_type?
             guard_clause.parent.source
+          elsif parent.or_type?
+            "return #{guard_clause.parent.source}"
```
Comment (KessaPassa) [inline lib/rubocop/cop/style/guard_clause.rb:166]: Do you know any better condition to switch on/off return?

Comment (daipom) [inline lib/rubocop/cop/style/guard_clause.rb:166]: Looks like we need `return` on the `and` case too.

I think this is not related to `and` or `or`.
It seems that the same problem exists in `and`.

The following two codes can have different results.

```ruby
if something
  work and return
else
  test
end
```

```ruby
work and return if something
test
```

It is the following case:

* `something` : `true`
* `work`: `false`

```ruby
if true
  false and return
else
  p "hoge"
end
=> (no output)
```

```ruby
false and return if true
p "hoge"
=> "hoge"
```

Comment (KessaPassa) [reply lib/rubocop/cop/style/guard_clause.rb:166]: @daipom Thank you for your comment !
I think your code is correct.
I'll check it.

Comment (KessaPassa) [reply lib/rubocop/cop/style/guard_clause.rb]: I fixed this commit.
d1f0a224fce9e6f351fbae5e709b95fb00e92489

Comment (daipom) [reply lib/rubocop/cop/style/guard_clause.rb]: Hmm, indeed it is...

Given this example, perhaps we should recommend this trailing-if-style (such as `work if cond`) only when we are sure that a `raise` or `return` will occur in the block.

It would be correct not to recommend this if it is unclear whether a `raise` or `return` will definitely occur in the block.

**Comment by bbatsov** [comment]
Can you summarize here what's the gist of the PR? I'm finding this a bit hard to understand from the code changes.

**Comment by KessaPassa** [comment]
summary: This PR fix is rubocop's suggested code to be equivalent to the original code.
Currently, rubocop warns as follows.

when nested guard clause with `or` operator.
before
```ruby
if something
  foo || raise('exception')
else
  ok
end
```

after
```ruby
foo || raise('exception') if something
ok
```
when 
`something: true`
`foo: true`
`ok` is executed.

when nested guard clause with `and` operator.
before
```ruby
if something
  foo and raise('exception')
else
  ok
end
```

after
```ruby
foo and raise('exception') if something
ok
```
when 
`something: true`
`foo: false`
`ok` is executed.

This PR changes rubocop's suggested code as follows.

when nested guard clause with `or` operator.
before
```ruby
if something
  foo || raise('exception')
else
  ok
end
```

after
```ruby
return foo || raise('exception') if something
ok
```
when 
`something: true`
`foo: true`
`ok` is not executed.

when nested guard clause with `and` operator.
before
```ruby
if something
  foo and raise('exception')
else
  ok
end
```

after
```ruby
return (foo and raise('exception')) if something
ok
```
when 
`something: true`
`foo: false`
`ok` is not executed.

@bbatsov Does this comment solve your question?

File: spec/rubocop/cop/style/guard_clause_spec.rb [hunk 2]
```diff
@@ -132,14 +132,20 @@ def func
           test
         end
       end
+
+      def func
```
Comment (bbatsov) [inline spec/rubocop/cop/style/guard_clause_spec.rb:136]: I don't think that's the right place for this bit of code, given the description of the test. You can add a separate test for it.

File: lib/rubocop/cop/style/guard_clause.rb [hunk 3]
```diff
@@ -226,9 +226,14 @@ def range_of_branch_to_remove(node, guard)
           branch.source_range
         end
 
-        def guard_clause_source(guard_clause)
+        def guard_clause_exiting_keyword(guard_clause)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/guard_clause.rb:229]: I don't see the point in renaming this. Frankly, I find the new name confusing, so if you still want to change it should be something that reflects better the purpose of the method.

**Comment by bbatsov** [comment]
Generally the `return` is not needed for one-line bodies method bodies, so in an ideal world this should be factored in when making such suggestions. Some of the tests were smaller they should have been.

**Comment by KessaPassa** [comment]
> Generally the return is not needed for one-line bodies method bodies

Did you mean this code:
```ruby
def func
  if something
    foo || raise('exception')
  end
end
```

should be autocorrected as:
```ruby
def func
  foo || raise('exception') if something
end
```

instead of:
```ruby
def func
  return foo || raise('exception') if something
end
```

?

if so this PR does not process this case.
